File: BaekJoon/16120.java
import java.io.*;
import java.util.*;

class main{
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String s = br.readLine();
		Stack<Character> stk = new Stack<>();
		for(int i = 0; i < s.length(); i++){
			char c = s.charAt(i);
			stk.push(c);
			if (stk.size() < 4) continue;
			while (stk.size() > 3){
				String tmp = "";
				for(int j = 0; j < 4; j++) tmp += stk.pop();
				if (tmp.equals("PAPP")) stk.push('P');
				else {
					for(int j = tmp.length() - 1; j >= 0; j--) stk.push(tmp.charAt(j));
					break;
				}
			}
		}
		System.out.println((stk.size() == 1 && stk.peek() == 'P' ? "PPAP" : "NP"));
	}
}

File: BaekJoon/16120.cpp
#include <iostream>
#include <stack>
using namespace std;
int main(){
	cin.tie(0)->sync_with_stdio(0);
	string s; cin >> s;
	stack<char> stk;
	for(char c : s){
		stk.push(c);
		if (stk.size() < 4) continue;
		while (stk.size() > 3){
			string tmp;
			for(int i = 0; i < 4; i++){
				tmp.push_back(stk.top());
				stk.pop();
			}
			if (tmp == "PAPP") stk.push('P');
			else {
				for(int i = tmp.length() - 1; i >= 0; i--) stk.push(tmp[i]);
				break;
			}
		}
	}
	cout << (stk.size() == 1 && stk.top() == 'P' ? "PPAP" : "NP");
}